django/contrib/admin/views/main.py

```diff
@@ -384,3 +391,135 @@ def url_for_result(self, result):
                                                self.opts.model_name),
                        args=(quote(pk),),
                        current_app=self.model_admin.admin_site.name)
+
+
+class ForeignKeyJsonPage(Page):
+    def has_next(self):
+        """
+        Return true until queryset is empty.
+
+        This avoids calling ``Queryset.count()`` which can be slow on large querysets.
+        """
+        return self.object_list.exists()
+
+
+class ForeignKeyJsonPaginator(Paginator):
+    """
+    A Django Paginator for big data sets.
+
+    Because users mostly use the first pages, it doesn't matter
+    if the page count is correct or not, as long as it's high enough.
+
+    Count() queries can take time, because they want to be exact.
+
+    So this paginator just doesn't count.
+    """
+
+    def _get_page(self, *args, **kwargs):
+        return ForeignKeyJsonPage(*args, **kwargs)
+
+    def validate_number(self, number):
+        """Validates the given 1-based page number but doesn't care about empty pages."""
+        try:
+            number = int(number)
+        except (TypeError, ValueError):
+            raise PageNotAnInteger('That page number is not an integer')
+        if number < 1:
+            raise EmptyPage('That page number is less than 1')
+        return number
+
+    @cached_property
+    def count(self):
+        if self.object_list.exists():
+            return sys.maxsize
+        return 0
+
+
+class ForeignKeyJsonView(BaseListView):
+    """
+    View that handles requests from `.AdminForeignKeyWidget`.
+
+    The view only supports HTTP's GET method.
+    """
+    paginator_class = ForeignKeyJsonPaginator
+    paginate_by = 20
+    admin_site = None
+
+    def get(self, request, *args, **kwargs):
+        """
+        Return JSON search result based on a field identifier and a search term.
+
+        Returns:
+             (django.http.JsonResponse)::
+                {
+                    'results': [
+                        {
+                            'text': "foo",
+                            'id': 123
+                        }
+                    ],
+                    'more': true
```

Comment:
With 4.0, Select2 recommended returning

```js
{
  'results': [],
  'pagination': {
    'more': true
  }
}
```

Where the `more` flag is nested within `pagination`. The old style, where it wasn't nested, is no longer supported.

This would prevent the need to have a `processResults` callback, which currently handles the mapping.